### PR TITLE
Generalize Add Zify to accept terms not just refs

### PR DESCRIPTION
**Pull request by SkySkimmer**
Help with https://github.com/coq/coq/issues/16404

This is like the "Specialize Zify" suggested there, but there is no need for a new command. So instead of `Specialize Zify FooOp FooInst v` we can do `Add Zify FooOp (FooInst v)`

Depends on https://github.com/coq/coq/pull/16802

### CODE REVIEW

**Comment by SkySkimmer** [comment]
@coqbot run full co

**Comment by SkySkimmer** [comment]
@coqbot run full ci

**Comment by SkySkimmer** [comment]
@coqbot run full ci

**Comment by jfehrle** [comment]
Changelog would be nice.

**Comment by SkySkimmer** [comment]
that's why there's the label

**Comment by SkySkimmer** [comment]
@coqbot run full ci

File: doc/changelog/04-tactics/16868-zify-term.rst [hunk 1]
```diff
@@ -0,0 +1,5 @@
+- **Added:** :cmd:`Add Zify` now accepts terms instead of references.
+  This is typically useful to do ``Add Zify InjTyp (inj_thing x)`` in
+  a section where ``x`` is a section variable without having to make
+  an intermediate definition for ``inj_thing`` (`#16868
```
Comment (jfehrle) [inline doc/changelog/04-tactics/16868-zify-term.rst:4]:
```suggestion
  This lets you avoid creating intermediate definitions for terms
  that include section variables (such as ``x`` in ``(inj_thing x)``).
  Now you can use ``Add Zify InjTyp (inj_thing x)`` instead (`#16868
```

WDYT?  A bit more readable if the benefit appears at the beginning of the sentence so the reader immediately knows why they would care.

**Review by jfehrle — COMMENTED** [review]
wording suggestions

File: doc/tools/docgram/orderedGrammar [hunk 2]
```diff
@@ -872,8 +872,8 @@ command: [
 | "Locate" reference
 | "Locate" "Term" reference
 | "Locate" "Module" qualid
+| "Add" "Zify" add_zify one_term      (* Micromega plugin *)
 | "Info" natural ltac_expr
-| "Add" "Zify" add_zify qualid      (* Micromega plugin *)
```
[the remark this answers is not in the document]
Comment (SkySkimmer) [reply doc/tools/docgram/orderedGrammar:876]: This is autogenerated, I don't know why the order changed

[the remark this answers is not in the document]
Comment (jfehrle) [reply doc/tools/docgram/orderedGrammar:876]: doc_grammar can't always match up the old and new productions, in which case it ends up in a semirandom spot.  Thus it's fine and at times necessary to reorder items in `orderedGrammar`.

**Comment by SkySkimmer** [comment]
Integrated @jfehrle comments

**Comment by SkySkimmer** [comment]
@coqbot run full ci

**Comment by Zimmi48** [comment]
@coq/micromega-maintainers It looks like this PR is just waiting to get reviewed.

**Review by ppedrot — CHANGES_REQUESTED** [review]
I think this is a regression. Allowing to store arbitrary terms in a map in a trick just waiting to blow at the face of users. As a matter of fact I have been pushing right in the other direction for features like hints. Like it or not, there is no universally agree upon equality function on terms, let alone a comparison function. Forcing the user to write an intermediate definition is the only sane semantics I can think of.

**Comment by SkySkimmer** [comment]
The thing that is put in the map is not the lemma `Inj_bv_Z w` which could be OK to have an intermediate definition for, it is the `bitvector w`. Forcing intermediate definitions for this is rather painful and also unstable as if they get unfolded lia breaks.

**Comment by fajb** [comment]
> I think this is a regression. 
Indeed, I remember `zify` used to take terms.

In this specific case, I guess the user does not wish to make an intermediate definition for say, int16, add16 etc.

What if the registered terms is checked to be an application of refs?
In that case, would the operations on terms be sufficiently well-behaved?
 =  : app_of_refs -> term -> bool 
<   : app_of_refs ->  term -> bool etc
That looks relatively lightweight to implement: a sanity check at registration time.

**Comment by ppedrot** [comment]
Recursive application of references is better, but there is still the (maybe not very relevant right now but that may be in the future) annoying issue of universe polymorphic definitions.

**Comment by fajb** [comment]
Could polymorphic definitions be ruled out ?
Are not they out-of-scope, anyway?  Given  `zify`  is mostly first-order logic.

**Comment by ppedrot** [comment]
> Could polymorphic definitions be ruled out?

Yes, but this is basically the confession that we're stuck forever with the current stdlib and the wealth of issues with monomorphic constants.

**Comment by ppedrot** [comment]
From what I gather from the code, we need some kind of dedicated pattern type + an efficient matching function. Storing terms in a map that are furthermore generated by calls to unsafe_to_constr is a literal recipe for an impending disaster.

**Comment by gares** [comment]
Not ready for 8.18, postponing to 8.19

**Pull request closed without merging**